### DISCUSSION TITLE
DAP Problems

### DISCUSSION THREAD

**Discussion by IlyasYOY** [mfussenegger/nvim-jdtls#412] [Q&A] [answered]
Hello! Thanks for making such a cool tool, it helps a lot!

I have a problem configuring DAP in my dotfiles.

I have 2 problems: 
1. No `JdtRefreshDebugConfigs ` after I [enable](https://github.com/IlyasYOY/dotfiles/blob/master/config/nvim/after/ftplugin/java.lua#L115) `com.microsoft.java.debug.plugin `.
2. Error after `vscode-java-test `. I attach screenshot with stacktrace. `JdtShowLogs` doesn't seem to help here. Error disappears when I remove `java-test` [extension](https://github.com/IlyasYOY/dotfiles/blob/master/config/nvim/after/ftplugin/java.lua#L122).

I have [called](https://github.com/IlyasYOY/dotfiles/blob/master/config/nvim/after/ftplugin/java.lua#L81) `setup_dap` function. Maybe I should [compile it from source](https://github.com/mfussenegger/nvim-jdtls/discussions/329)? 🤔 

Java version output: 
```
openjdk version "17.0.4" 2022-07-19 LTS
OpenJDK Runtime Environment (build 17.0.4+8-LTS)
OpenJDK 64-Bit Server VM (build 17.0.4+8-LTS, mixed mode, sharing)
```

<img width="1536" alt="image" src="https://user-images.githubusercontent.com/18665585/216123592-d007b425-e0e3-4b9d-8085-8e29e1c8e1c0.png">

**Comment by mfussenegger** [comment, marked as answer]
Looks like your bundles definition is wrong. `vim.fn.glob` returns a string and `bundles` must be a list with the path to a `jar` per item.

You could split the result of `vim.fn.glob` by newlines:

```lua
 init_options = {
   bundles = vim.split(
    vim.fn.glob(
          Path.path.home
              .. "/.local/share/nvim/mason/packages/"
              .. "java-debug-adapter/extension/"
              .. "server/com.microsoft.java.debug.plugin-*.jar",
          1
    ),
    '\n'
  )
}
```

**Reply by IlyasYOY** [reply]
I added split as you suggested & now I receive the exception on screenshot. 

It says `java.lang. ClassCastException: class java.til.ArrayList cannot be cast to class java.lang.String` ([here](https://github.com/eclipse/eclipse.jdt.ls/blob/29a809c2aa6b11d358c3c6ce6b17b4a4ce010bc7/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/BundleUtils.java#L107)). I guess `bundles` are expected to be strings, that is strange. Everywhere (in docs and examples) I see that bundles` is `List` of `String`, but LSP doesn't accept that. When I pass it as a `String` I do not get DAP setup (no `JdtRefreshDebugConfigs` command)

I have JDTL v1.19.0 installed. 

PS. LSP doesn't throw an exception when I don't split string 🤔 
But still don't get `JdtRefreshDebugConfigs `


```lua
    init_options = {
        bundles = {
            G_P(
                vim.fn.glob(
                    get_install_path_for "java-debug-adapter"
                        .. "/extension/server/"
                        .. "com.microsoft.java.debug.plugin-*.jar",
                    1
                )
            ),
            G_P(
                vim.fn.glob(
                    get_install_path_for "java-test"
                        .. "/extension/server/"
                        .. "*.jar",
                    1
                )
            ),
        },
    },
```

Server works fine.  Output I receive: 

```
"/Users/ilyasyoy/.local/share/nvim/mason/packages/java-debug-adapter/extension/server/com.microsoft.java.debug.plugin-0.44.0.jar"
"/Users/ilyasyoy/.local/share/nvim/mason/packages/java-test/extension/server/com.microsoft.java.test.plugin-0.37.1.jar\n/Users/ilyasyoy/.local/share/nvim/mason/p
ackages/java-test/extension/server/com.microsoft.java.test.runner-jar-with-dependencies.jar\n/Users/ilyasyoy/.local/share/nvim/mason/packages/java-test/extension
/server/junit-jupiter-api_5.9.0.jar\n/Users/ilyasyoy/.local/share/nvim/mason/packages/java-test/extension/server/junit-jupiter-engine_5.9.0.jar\n/Users/ilyasyoy/
.local/share/nvim/mason/packages/java-test/extension/server/junit-jupiter-migrationsupport_5.9.0.jar\n/Users/ilyasyoy/.local/share/nvim/mason/packages/java-test/
extension/server/junit-jupiter-params_5.9.0.jar\n/Users/ilyasyoy/.local/share/nvim/mason/packages/java-test/extension/server/junit-platform-commons_1.9.0.jar\n/U
sers/ilyasyoy/.local/share/nvim/mason/packages/java-test/extension/server/junit-platform-engine_1.9.0.jar\n/Users/ilyasyoy/.local/share/nvim/mason/packages/java-
test/extension/server/junit-platform-launcher_1.9.0.jar\n/Users/ilyasyoy/.local/share/nvim/mason/packages/java-test/extension/server/junit-platform-runner_1.9.0.
jar\n/Users/ilyasyoy/.local/share/nvim/mason/packages/java-test/extension/server/junit-platform-suite-api_1.9.0.jar\n/Users/ilyasyoy/.local/share/nvim/mason/pack
ages/java-test/extension/server/junit-platform-suite-commons_1.9.0.jar\n/Users/ilyasyoy/.local/share/nvim/mason/packages/java-test/extension/server/junit-platfor
m-suite-engine_1.9.0.jar\n/Users/ilyasyoy/.local/share/nvim/mason/packages/java-test/extension/server/junit-vintage-engine_5.9.0.jar\n/Users/ilyasyoy/.local/shar
e/nvim/mason/packages/java-test/extension/server/org.apiguardian.api_1.1.2.jar\n/Users/ilyasyoy/.local/share/nvim/mason/packages/java-test/extension/server/org.e
clipse.jdt.junit4.runtime_1.3.0.v20220609-1843.jar\n/Users/ilyasyoy/.local/share/nvim/mason/packages/java-test/extension/server/org.eclipse.jdt.junit5.runtime_1.
1.0.v20220715-1030.jar\n/Users/ilyasyoy/.local/share/nvim/mason/packages/java-test/extension/server/org.opentest4j_1.2.0.v20190826-0900.jar"
```

But when I split everything falls apart.

- - -

Code like this doesn't cause an exception, but DAP doesnt work

```lua
    init_options = {
        bundles = G_P(coredor.array_flat_map({ -- My flat map, works fine
            G_P(
                vim.fn.glob(
                    get_install_path_for "java-debug-adapter"
                        .. "/extension/server/"
                        .. "com.microsoft.java.debug.plugin-*.jar",
                    1
                )
            ),
            G_P(
                vim.fn.glob(
                    get_install_path_for "java-test"
                        .. "/extension/server/"
                        .. "*.jar",
                    1
                )
            ),
        }, function(str)
            return coredor.string_split(str, "\n") -- my split (has plain = true hardcoded)
        end)),
    },
```

**Reply by mfussenegger** [reply]
`bundles` requires a list of strings. I missed initially that you've _two_ vim.fn.glob calls. If you split on both you'll get a nested list, and that doesn't work.

You'll need to flatten it out, e.g. with something like this:

```lua
local java_debug = get_install_path_for("java-debug-adapter") .. "/extension/server/com.microsoft.java.debug.plugin-*.jar"
local java_test = get_install_path_for("java-debug-adapter") .. "/extension/server/*.jar"
local bundles = {}
vim.list_extend(bundles, vim.split(vim.fn.glob(java_debug, 1), '\n'))
vim.list_extend(bundles, vim.split(vim.fn.glob(java_test, 1), '\n'))
init_options = {
  bundles = bundles
}
```

**Reply by IlyasYOY** [reply]
Yeah, this way it LSP does not fail. But it also seems line DAP is not been configured by `setup_dap` ([here](https://github.com/IlyasYOY/dotfiles/blob/master/config/nvim/after/ftplugin/java.lua#L83))

**Reply by mfussenegger** [reply]
You need to call `jdtls.setup.add_commands()` after the `setup_dap` call to get `JdtRefreshDebugConfigs`

**Reply by IlyasYOY** [reply]
oh, sure! thanks 🙏 
now everything works like a charm! 

Maybe I could help adding mason examples to the `README.md`? 
Or full config in the separate file, so a newcomer would be able to reuse them.

**Reply by mfussenegger** [reply]
You could add something to https://github.com/mfussenegger/nvim-jdtls/wiki/Sample-Configurations

I'm not going to add Mason specific examples to the readme but stick to the lowest common denominator

**Reply by IlyasYOY** [reply]
Hello, I created standalone config with Mason: [Simple JDTLS config for NeoVim](https://gist.github.com/IlyasYOY/f0105354306b5b37fa2d6805512155d0)
